src/main/java/com/google/sps/Scheduler.java

```diff
@@ -1,15 +1,134 @@
 package com.google.sps;
 
-import java.util.Collection;
-
 import com.google.sps.data.*;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
 
 /**
  * Static class for permuting schedules given a list of courses.
  */
 public final class Scheduler {
-    public Collection<Schedule> generateSchedules(Collection<Course> courses, Invariants invariants){
-        //TODO: This
-        return null;
+  /**
+   * Generates all valid schedules given a list of courses and invariant
+   * constraints.
+   * 
+   * @param courses    The list of courses to generate the schedules from.
+   * @param invariants The invariants that constrain the valid schedules
+   * @return A list of valid schedules.
+   */
+  public List<Schedule> generateSchedules(List<Course> courses, Invariants invariants) {
+    List<Course> requiredCourses = new ArrayList<>();
+    List<Course> nonRequiredCourses = new ArrayList<>();
+
+    for (Course course: courses) {
+      if (course.isRequired()) {
+        requiredCourses.add(course);
+      } else {
+        nonRequiredCourses.add(course);
+      }
+    }
+
+    // Generate all possible valid Schedules that include only required courses.
+    float reqCredits = totalCredits(requiredCourses);
+    Set<List<Course>> requiredSet = new HashSet<>();
+
+    generateSchedulesHelper(requiredCourses, new ArrayList<>(), 
+        new Invariants(reqCredits, reqCredits), requiredSet);
+    
+    // Using those Schedules as a building block, generate all valid schedules
+    Set<List<Course>> scheduleSet = new HashSet<>();
+    for (List<Course> validCourseList : requiredSet) {
+      generateSchedulesHelper(nonRequiredCourses, validCourseList, invariants, scheduleSet);
+    }
+
+    List<Schedule> schedules = new ArrayList<>();
+    for (List<Course> courseList: scheduleSet) {
+      schedules.add(new Schedule(courseList));
+    }
+
+    return schedules;
+  }
+
+  /**
+   * Recursively iterates through a list of courses and adds schedules to the
+   * given set.
+   * 
+   * @param courses     The list of courses to iterate through
+   * @param courseList  The current course list to be added/operated on
+   * @param invariants  The invariants that constrain the valid schedules
+   * @param courseLists The set of course lists that is being generated/added to.
+   */
+  private void generateSchedulesHelper(List<Course> courses, List<Course> courseList, 
+      Invariants invariants, Set<List<Course>> courseLists) {
```

Comment:
Similar to the comment above, `courseLists` is hard to understand here. Alternative: `generatedCourseLists`. It's not that great of a suggestion but seems like a step in the right direction.